### ISSUE TITLE
Stack Cloudformation Names are Nondeterministic

### ISSUE THREAD

**Issue by joekiller** [aws-amplify/amplify-cli#12640]
### How did you install the Amplify CLI?

npm 8.19.3

### If applicable, what version of Node.js are you using?

v16.19.1

### Amplify CLI Version

11.1.1

### What operating system are you using?

Mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No manual changes made

### Describe the bug

During the creation of a stack the stackname created is nondeterministic which makes it impossible to determine the stack name and pre-push activities which need stack name information.

The bug is in the [cloudformation initializer].

[cloudformation initializer]:https://github.com/aws-amplify/amplify-cli/blob/v12.0.0/packages/amplify-provider-awscloudformation/src/initializer.ts#L60-L62

### Expected behavior

I should be able to determine what a stack name will be based on context within the build without relying on a random timestamp or UUID string.

### Reproduction steps

1. Create an amplify application without any environments and hook up CI builds.
2. Push a change to a CI environment.
3. A random backend environment name is chosen and the stack created has a random id in it.

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Before submitting, please confirm:

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

**Comment by joekiller** [comment]
@josefaidt hello! 

I will explain how the random environment name is created. Please understand that "the random environment name" is not the non-deterministic part of the problem. Although it is random in a sense, it is available to the build in a predicable way allow for a reproducible and predictable build. This random name is "deterministic" in my view because it's in the pre-hook and environment variables at push. With that preamble, below is how a random backend environment name is chosen.

Enable [pattern based branch deployment]. Create a new branch with a name that'll be deployed and push the branch. Use a branch name like feature/12640 to ensure the provision step will provide a random name. The last line of the provision step will mention the new environment name.

After the provision step is complete, the environment is named however a stack hasn't been deployed so the stackname will be {appName}-{envName}-{random uuid if circleci else hhmms timestamp}.

We use envName pre-push.js [parameters] input but have found that during the initial CI push, where the environment key doesn't exist, the parameter data structure is different than what is documented. I plan to file a separate bug regarding the difference in parameter value shapes. Regardless, we do stuff in this hook, even if nothing exists, and not being able to determine the stackname is breaking our ability to use the pattern based deployment integration.

Because the amplify-metadata.json hasn't been written and the stack doesn't exist the name at that point is non-deterministic due to the code I mentioned.


[pattern based branch deployment]:https://docs.aws.amazon.com/amplify/latest/userguide/pattern-based-feature-branch-deployments.html
[parameters]:https://docs.amplify.aws/cli/project/command-hooks/#how-to-name-command-hook-scripts

**Comment by joekiller** [comment]
Hello, I said the provision step for the random name but it is the last step of the checkout step. `Random environment name XYZ generated`

**Comment by josefaidt** [comment]
Hey @joekiller thanks for posting those details! I started on a reproduction and immediately ran into a different error: 
![image](https://github.com/aws-amplify/amplify-cli/assets/5033303/367c0807-649c-480c-b06a-18baa118f5d9)

After re-reading your comment I'm curious, are you using the pre-push hook to setup the SSM Parameters required for the new env?

> it is the last step of the checkout step. Random environment name XYZ generated

Ah I see this in the clone/checkout step:

![image](https://github.com/aws-amplify/amplify-cli/assets/5033303/f2116456-07e7-45dd-a5cb-8384f48cb629)

For the random env naming I will need to transfer this to our Hosting repo, however I'd like to understand the intention with the hook as well as track the SSM Parameter issue separately (will link here soon)

**Comment by joekiller** [comment]
@josefaidt I do not think this should be transferred because the random environment name is fine. The environment name is known and in the build environment as appropriate. 

At amplify push the cli relies on the amplify-meta.json to resolve the env value. When an environment is already present the cli has access to the value. Our problem is that given that information we cannot predict the stackname when pushing the first time and the push also creates the environment at the same time.

We have a pre-push.js script which makes some resources available given the stackname in s3 that fails to deploy when pushing to an environment that doesn't exist because the stackname itself has a suffix that is non-deterministic. The environment name is known as that point but the stack name is unpredictable because it is resolved after the hook runs. 

If the stack naming was deterministic then our script would be able to determine the stackname given the environment and application context even with the amplify-meta.json file not being present yet. The pre-push.js cannot predict the name due to the amplify cli making all new stack names include a suffix that it either a  uuid or date string generated at that moment.

**Comment by joekiller** [comment]
I'll give you a commit on your test repo to provoke the issue tomorrow.

**Comment by joekiller** [comment]
My colleague was implementing the [following workaround](https://github.com/aws-amplify/amplify-cli/issues/9055#issuecomment-1471518533) when we noticed it wouldn't work for new environments due to the amplify.envirinment.envName values not being populated in the cli for the initial push while invoking the pre-push script.

**Comment by joekiller** [comment]
Prepopulating ssm vars is kinda a cool idea too. If they were tied to the stackname the same problem would apply that while the project name and environment are known values, the parent stack will still have some random suffix generated at deploy time.

**Comment by joekiller** [comment]
Seems this applies mainly during an init --forcePush

**Comment by joekiller** [comment]
Finally, note on an initial forcePush that the args that the pre-push hook gets doesn't contain an environment key.

According to the code the [DataParameter] might not have environment keys. So the docs about the [data parameter structure](https://docs.amplify.aws/cli/project/command-hooks/#undefined-parameter-structure) should be updated to reflect that there can be optionality/undefined in the keys.


[DataParameter]:https://github.com/aws-amplify/amplify-cli/blob/v12.0.0/packages/amplify-cli-core/src/hooks/hooksTypes.ts#L25-L33

First init forcePush:
>{"data":{"amplify":{"argv":["/root/.amplify/bin/amplify","/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/bin/amplify","init","--amplify","{\"envName\":\"trixy\",\"appId\":\"abc123\"}","--providers","{\"awscloudformation\":{\"configLevel\":\"project\",\"useProfile\":true,\"profileName\":\"default\"}}","--codegen","{\"generateCode\":false,\"generateDocs\":false}","--yes","--categories","{\"auth\":{\"userPoolId\":\"us-east-abc123\",\"webClientId\":\"abc123\",\"nativeClientId\":\"abc123\"}}","--forcePush"],"version":"10.5.2"}}}

Subsequent init forcePush:

> {"data":{"amplify":{"environment":{"projectPath":"/codebuild/output/src878451092/src/dcon-next","envName":"trixy"},"argv":["/root/.amplify/bin/amplify","/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/bin/amplify","init","--amplify","{\"envName\":\"trixy\",\"appId\":\"abc123\"}","--providers","{\"awscloudformation\":{\"configLevel\":\"project\",\"useProfile\":true,\"profileName\":\"default\"}}","--codegen","{\"generateCode\":false,\"generateDocs\":false}","--yes","--categories","{\"auth\":{\"userPoolId\":\"us-east-abc123\",\"webClientId\":\"abc123\",\"nativeClientId\":\"abc123\"}}","--forcePush"],"version":"10.5.2"}}}

**Comment by josefaidt** [comment]
Hey @joekiller thanks for diving into this and providing those details! I'm going to mark this as a feature request to assess the general ask of deterministic stack names, perhaps through some sort of naming convention project override.

> Prepopulating ssm vars is kinda a cool idea too. If they were tied to the stackname the same problem would apply that while the project name and environment are known values, the parent stack will still have some random suffix generated at deploy time.

This will work, and the parameters are tied to the Amplify app ID and environment name. More information on the naming convention can be found here https://docs.amplify.aws/cli/reference/ssm-parameter-store/#manually-creating-parameters

> Seems this applies mainly during an init --forcePush

I do see where the env parameters check is running, then the prepush hook invocation (I'll add this to the other issue as well, thanks for opening that separately!):

https://github.com/aws-amplify/amplify-cli/blob/dev/packages/amplify-cli/src/commands/init.ts#L37

**Comment by joekiller** [comment]
Thanks @josefaidt a simple env option is likely enough.
